### PR TITLE
[FL-2744] SPI Mem Manager C port

### PR DESCRIPTION
**Pull request by drunkbatya**
# What's new

- new application - SPI Mem Manager, it allows reading, writing, comparing SPI chip dumps and erasing SPI chips

# Verification 

- try `./fbt launch_app APPSRC=applications/plugins/spi_mem_manager`, connect SPI chip to external SPI bus and test it. all functions should work

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_chip.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#include "spi_mem_chip.h"
+
+const SPIMemChipVendorName SPIMemChipVendorNames[] = {
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_chip.c]: We use snake case for variables and function names, so `spi_mem_chip_vendor_names`.

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+typedef void (*SPIMemWorkerChipDetectCallback)(void* context, SPIMemWorkerChipDetectResult result);
+
+typedef struct {
+    SPIMemChip* chip_info;
+    SPIMemWorkerMode mode_index;
+    SPIMemWorkerChipDetectCallback chip_detect_cb;
+    void* cb_ctx;
+    FuriThread* thread;
+} SPIMemWorker;
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.h]: Do not export internal structures to entity API.

**Comment by AkechiShiro** [comment]
Hey, I'd like to ask is this PR far from being finished, I've seen lots of commits being made, I'd like to know if there is any chance of it being planned soon for a near release?

**Comment by drunkbatya** [comment]
> Hey, I'd like to ask is this PR far from being finished, I've seen lots of commits being made, I'd like to know if there is any chance of it being planned soon for a near release?

Hi! Thanks for interesting this app!
I think the first version will be included at the next firmware release. Before it we need to add support of most popular SPI chips, and do some tests.

**Comment by AkechiShiro** [comment]
@drunkbatya Thanks, I'm aware that this might still be experimental, I hope I may be able to help with testing. 

Also, about adding more SPI flash chips, this might be of use, I'm not sure if [all the data from there could help]( https://www.flashrom.org/Supported_hardware#Supported_flash_chips), then a tracking issue could be created with checkboxes for each SPI flash that is planned to be supported, the FP Zero can support, I believe 3.3V/5V.

Question: 1.8V is not doable ? I don't know if the GPIOs can be set to 1.8V. Maybe later, it could be achieved? 

Once this draft PR merges into a release, best would be to be able to test the ones that are already supported (Winbonds chips I believe) first and see if reading/writing works alright (compared to a Raspberry Pi or any other device capable of read/writes to SPI chips).

A future feature, that could be worked on, once SPI reading/writing works alright is this : https://www.macronix.com/Lists/ApplicationNote/Attachments/1870/AN114v1-SFDP%20Introduction.pdf

File: applications/plugins/application.fam [hunk 3]
```diff
@@ -5,5 +5,7 @@ App(
     provides=[
         "music_player",
         "snake_game",
+        "bt_hid",
```
Comment (DrZlo13) [inline applications/plugins/application.fam]: ?

File: applications/plugins/spi_mem_manager/application.fam [hunk 4]
```diff
@@ -0,0 +1,17 @@
+App(
+    appid="spi_mem_manager",
+    name="SPI Mem Manager",
+    apptype=FlipperAppType.EXTERNAL,
+    entry_point="spi_mem_app",
+    requires=["gui"],
+    stack_size=1 * 2048,
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/application.fam:7]: check stack via "ps" in cli

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_chip_i.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+} SPIMemChipVendorName;
+
+struct SPIMemChip {
+    SPIMemChipVendor vendor_enum;
+    uint8_t vendor_id;
+    const char* model_name;
+    size_t size;
+    SPIMemChipWriteMode write_mode;
+    uint8_t type_id;
+    uint8_t capacity_id;
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_chip_i.h:77]: group and align uint8_t values

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.c [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+static int32_t spi_mem_worker_thread(void* thread_context);
+
+SPIMemWorker* spi_mem_worker_alloc() {
+    SPIMemWorker* worker = malloc(sizeof(SPIMemWorker));
+    worker->callback = NULL;
+    worker->thread = furi_thread_alloc();
+    worker->mode_index = SPIMemWorkerModeIdle;
+    furi_thread_set_name(worker->thread, "SPIMemWorker");
+    // furi_thread_set_priority(worker->thread, FuriThreadPriorityHigh);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.c]: Remove commented code

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.c [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    furi_thread_join(worker->thread);
+}
+
+void spi_mem_worker_chip_detect_start(
+    SPIMemChip* chip_info,
+    found_chips_t* found_chips,
+    SPIMemWorker* worker,
+    SPIMemWorkerCallback callback,
+    void* context) {
+    furi_assert(worker->mode_index == SPIMemWorkerModeIdle);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker.c]: furi_check?

File: applications/plugins/spi_mem_manager/spi_mem_files.c [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+size_t spi_mem_file_get_size(SPIMemApp* app) {
+    size_t size = 0;
+    bool success = false;
+    app->file = storage_file_alloc(app->storage);
+    do {
+        if(!storage_file_open(
+               app->file, furi_string_get_cstr(app->file_path), FSAM_READ, FSOM_OPEN_EXISTING))
+            break;
+        size = storage_file_size(app->file);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/spi_mem_files.c]: storage_common_stat

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker_modes.c [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+        SPIMemChipStatus chip_status = spi_mem_tools_get_chip_status(worker->chip_info);
+        if(chip_status == SPIMemChipStatusError) return false;
+        if(chip_status == SPIMemChipStatusBusy) continue;
+        return true;
+    }
+}
+
+static size_t spi_mem_worker_modes_get_total_size(SPIMemWorker* worker) {
+    size_t chip_size = spi_mem_chip_get_size(worker->chip_info);
+    size_t file_size = spi_mem_file_get_size(worker->cb_ctx);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker_modes.c:39]: Do not use callback context in non-callback functions

File: applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker_modes.c [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+    size_t total_size = chip_size;
+    if(chip_size > file_size) total_size = file_size;
+    return total_size;
+}
+
+// ChipDetect
+static void spi_mem_worker_chip_detect_process(SPIMemWorker* worker) {
+    SPIMemCustomEventWorker event;
+    while(!spi_mem_tools_read_chip_info(worker->chip_info)) {
+        furi_delay_tick(10);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/lib/spi/spi_mem_worker_modes.c]: Comment please

File: applications/plugins/spi_mem_manager/scenes/spi_mem_scene_read_filename.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+#include "../spi_mem_files.h"
+
+void spi_mem_scene_read_filename_view_result_callback(void* context) {
+    SPIMemApp* app = context;
+    view_dispatcher_send_custom_event(app->view_dispatcher, SPIMemCustomEventTextEditResult);
+}
+
+void spi_mem_scene_read_set_random_filename(SPIMemApp* app) {
+    if(furi_string_end_with(app->file_path, SPI_MEM_FILE_EXTENSION)) {
+        size_t filename_start = furi_string_search_rchar(app->file_path, '/');
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/scenes/spi_mem_scene_read_filename.c:11]: path_extract_dirname

File: applications/plugins/spi_mem_manager/scenes/spi_mem_scene_verify.c [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+bool spi_mem_scene_verify_on_event(void* context, SceneManagerEvent event) {
+    SPIMemApp* app = context;
+    UNUSED(app);
+    bool success = false;
+    if(event.type == SceneManagerEventTypeBack) {
+        success = true;
+    } else if(event.type == SceneManagerEventTypeCustom) {
+        success = true;
+        if(event.event == SPIMemCustomEventViewVerifySkip) {
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/scenes/spi_mem_scene_verify.c:38]: switch?

File: applications/plugins/spi_mem_manager/views/spi_mem_view_progress.c [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+    }
+    return total_size;
+}
+
+void spi_mem_view_progress_inc_progress(SPIMemProgressView* app) {
+    with_view_model(
+        app->view,
+        SPIMemProgressViewModel * model,
+        {
+            size_t total_size = spi_mem_view_progress_set_total_size(model);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/views/spi_mem_view_progress.c:208]: if(total_size == 0) total_size = 1;

File: applications/plugins/spi_mem_manager/views/spi_mem_view_progress.h [hunk 14]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+#include <gui/view.h>
+
+typedef struct SPIMemProgressView SPIMemProgressView;
+typedef void (*SPIMemProgressViewCallback)(void* context);
+
+View* spi_mem_view_progress_get_view(SPIMemProgressView* app);
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/views/spi_mem_view_progress.h:7]: Documentate functions

File: applications/plugins/spi_mem_manager/tools/chiplist.xml [hunk 15]
```diff
@@ -0,0 +1,979 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (DrZlo13) [inline applications/plugins/spi_mem_manager/tools/chiplist.xml]: Add authors, links and license

**Review by DrZlo13 — CHANGES_REQUESTED** [review]

File: applications/plugins/spi_mem_manager/views/spi_mem_view_progress.h (hunk 14 above)
Comment (skotopes) [reply applications/plugins/spi_mem_manager/views/spi_mem_view_progress.h:7]: ?

**Review by skotopes — DISMISSED** [review]

File: applications/plugins/application.fam (hunk 3 above)
Comment (DrZlo13) [reply applications/plugins/application.fam]: Still looks like a merge error.

**Review by DrZlo13 — CHANGES_REQUESTED** [review]

**Review by skotopes — APPROVED** [review]

**Review by hedger — APPROVED** [review]

**Review by DrZlo13 — APPROVED** [review]

**Pull request merged**